### DISCUSSION TITLE
flate: read data with partial flush

### DISCUSSION THREAD

**Discussion by vajexal** [klauspost/compress#993] [General]
problem: data written with zlib's `Z_PARTIAL_FLUSH` can't be read because decompressor returns only after `Z_SYNC_FLUSH` block or data end (not taking into account internal buffer flush). Maybe flush mode parameter could be added to decompressor? Using this parameter we could control [finishBlock](https://github.com/klauspost/compress/blob/3868468e621bb63b26416989999e274a9e4347dd/flate/inflate.go#L655) method to return after each block. Something like

```go
func (f *decompressor) finishBlock() {
	if f.final {
		if f.dict.availRead() > 0 {
			f.toRead = f.dict.readFlush()
		}
		
		f.err = io.EOF
	} else if f.flushMode == PARTIALPARTIAL_FLUSH && f.dict.availRead() > 0 {
		f.toRead = f.dict.readFlush()
	}

	f.step = nextBlock
}
```

New constructor (something like `NewPartialBlockReader`) will be needed

code to test
```go
package main

import (
	"bytes"
	"encoding/base64"
	"fmt"

	"github.com/klauspost/compress/flate"
)

func main() {
	data, err := base64.StdEncoding.DecodeString("ykjNycnXUSjPL8pJAQg=")
	if err != nil {
		panic(err)
	}

	r := flate.NewReader(bytes.NewReader(data))
	rb := make([]byte, 32)
	n, err := r.Read(rb)
	if err != nil {
		panic(err)
	}

	expected := "hello, world"
	actual := string(rb[:n])
	if expected != actual {
		panic(fmt.Sprintf("expected: %v, got: %v", expected, actual))
	}
}
```

**Comment by klauspost** [comment]
@vajexal Ah, so it slaps on an empty block with predefined table. But yeah, we will not have all the 10 bits, so it wouldn't be something we could do automatically.

Actually adding the option is a bit more annoying, since they are returning an interface. I am not a fan of "yet another New only for this", and adding an interface seems annoying to use. I think a `func NewReaderOpts(w, ReaderOptions{})` is the best I can come up with.

**Reply by vajexal** [reply]
if you ok with adding new constructor (`NewReaderOpts`), i can make a pr

**Reply by klauspost** [reply]
@vajexal Sure. A `Dict []byte` is the only other property that comes to mind. In the future a decode window max could maybe be relevant.
